### ISSUE TITLE
Singleton

### ISSUE THREAD

**Issue by MohamedAnwarAlhamed** [MohamedAnwarAlhamed/CSharp_Design_Patterns#22]
The Singleton Design Pattern is a creational design pattern that ensures a class has only one instance and provides a global point of access to that instance. It is useful when exactly one object is needed to coordinate actions across a system, such as managing shared resources, database connections, or logging.

Intent
-
Ensure a class has only one instance throughout the application.
Provide global access to that instance.

Problem
-
In some cases, you may need a single point of control or a shared resource that should not be duplicated. For instance, creating multiple instances of a configuration manager or a database connection object could lead to issues like inconsistent behavior, resource contention, or unnecessary overhead.

For example, if you have a logging system, it would be inefficient to have multiple loggers operating independently. You only need one logger to maintain consistent logs throughout the application.

Solution
-
The Singleton pattern solves this problem by:
- Restricting object creation: It ensures that only one instance of the class can be created.
- Providing global access: This single instance is accessed globally through a static method.

Structure:
- Private Constructor: Prevents clients from creating new instances using the new keyword.
- Static Instance Variable: Holds the single instance of the class.
- Static Method: Provides global access to the single instance.

Create the Singleton Class
```c#
// Singleton class
public class Singleton
{
    // Private static variable to hold the single instance
    private static Singleton instance;

    // Private constructor to prevent external instantiation
    private Singleton() 
    {
        Console.WriteLine("Singleton Instance Created.");
    }

    // Public static method to provide global access to the instance
    public static Singleton GetInstance()
    {
        // Lazy initialization: Create the instance if it doesn't exist
        if (instance == null)
        {
            instance = new Singleton();
        }
        return instance;
    }

    public void DoSomething()
    {
        Console.WriteLine("Doing something with the Singleton instance.");
    }
}

```

 Using the Singleton
```c#
class Program
{
    static void Main(string[] args)
    {
        // Get the Singleton instance and call a method on it
        Singleton s1 = Singleton.GetInstance();
        s1.DoSomething();  // Output: Singleton Instance Created.
                           // Output: Doing something with the Singleton instance.

        // Try to get another instance (it will return the same instance)
        Singleton s2 = Singleton.GetInstance();
        s2.DoSomething();  // Output: Doing something with the Singleton instance.

        // Check if both instances are the same
        Console.WriteLine(s1 == s2);  // Output: True
    }
}

```
Thread Safety
```c#
public class ThreadSafeSingleton
{
    private static ThreadSafeSingleton instance;
    private static readonly object lockObject = new object();

    // Private constructor
    private ThreadSafeSingleton() { }

    // Static method to get the singleton instance
    public static ThreadSafeSingleton GetInstance()
    {
        if (instance == null)
        {
            // Lock to ensure only one thread can enter this block
            lock (lockObject)
            {
                // Double-check if the instance is still null
                if (instance == null)
                {
                    instance = new ThreadSafeSingleton();
                }
            }
        }
        return instance;
    }
}

```

Eager Initialization
```c#
public class EagerSingleton
{
    private static readonly EagerSingleton instance = new EagerSingleton();

    private EagerSingleton() { }

    public static EagerSingleton GetInstance()
    {
        return instance;
    }
}

```

Lazy Initialization with Lazy<T> (C# Specific)
```c#
public class LazySingleton
{
    private static readonly Lazy<LazySingleton> instance = new Lazy<LazySingleton>(() => new LazySingleton());

    private LazySingleton() { }

    public static LazySingleton GetInstance()
    {
        return instance.Value;
    }
}

```

Advantages of the Singleton Pattern
-
- Controlled Access: Only one instance of the class exists, controlling access to a shared resource.
- Lazy Initialization: The instance is only created when needed, saving resources.
- Prevents Resource Conflicts: Avoids issues caused by multiple instances of a shared resource, such as database connections or configurations.
- Global Access: The instance can be accessed globally from anywhere in the application.

Disadvantages of the Singleton Pattern
-
- Global State: A singleton can act as a global variable, which can introduce hidden dependencies and make the system harder to understand and test.
- Testing Challenges: Singletons can be difficult to test because they introduce global state, making it harder to isolate individual tests.
- Concurrent Issues: If not implemented properly in multithreaded environments, it can result in multiple instances being created.

When to Use the Singleton Pattern
-
- When you need exactly one instance of a class to control shared resources.
- When you need global access to that instance, such as a logger, configuration manager, or connection pool.
- When creating single instances of expensive-to-create objects like databases or configuration settings makes sense.

Avoiding Overuse
-
While Singleton can be very useful in certain cases, overusing it can introduce problems like tight coupling and global state management. It's essential to use Singleton only when necessary and avoid turning it into a "default" pattern for all shared resources.

In summary, the Singleton Design Pattern is a powerful tool for ensuring that a class has only one instance, providing a global access point, and managing shared resources efficiently. It simplifies control over instance creation and promotes reuse while ensuring consistency across the system. However, it should be used carefully to avoid introducing global state and other potential issues.
